File: 백준/Silver/2304. 창고 다각형/창고 다각형.py
n = int(input())
arr = [list(map(int,input().split(" "))) for _ in range(n)]
arr.sort()

mx_idx = mx =0
for i in range(len(arr)):
    if mx < arr[i][1]:
        mx_idx = i
        mx = arr[i][1]


result = 0
left = arr[0]
for i in range(1,mx_idx+1):
    if arr[i][1] >= left[1]:
        result += left[1] * (arr[i][0] - left[0])
        left = arr[i]


right = arr[-1]
for i in range(len(arr)-2, mx_idx-1, -1):
    if (arr[i][1] >= right[1]):
        result += right[1] * (- arr[i][0] + right[0])
        right = arr[i]

print(result+mx)

File: 백준/Silver/2304. 창고 다각형/창고 다각형.java
import javax.management.AttributeList;
import javax.swing.text.html.parser.Parser;
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.InputStreamReader;
import java.util.*;
import java.util.stream.Collectors;

public class Main {

    public static void  main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader((new InputStreamReader(System.in)));
        StringTokenizer st ;

        int N = Integer.parseInt(br.readLine());
        int[][] array = new int[N][2];

        int mxIndex = 0;
        int mx = 0;

        int value;
        int index;
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            index = Integer.parseInt(st.nextToken());
            value = Integer.parseInt(st.nextToken());

            array[i][0] = index;
            array[i][1] = value;

            if(mx <= value){
                mx = value;
                mxIndex = index;
            }
        }

        Arrays.sort(array,(a,b)->{
            return a[0] - b[0];
        });


        int finalMxIndex = mxIndex;
        List<int[]> left = Arrays.stream(array).filter(a -> {
            return a[0] <=  finalMxIndex;
        }).collect(Collectors.toList());


        List<int[]> right = new ArrayList<>(Arrays.stream(array).filter(a -> {
            return a[0] >= finalMxIndex;
        }).collect(Collectors.toList()));

        right.sort((a,b)->{
            return b[0] - a[0];
        });


        int prev = 0;
        int result = 0;

        for (int i = 1; i < left.size(); i++) {
            if (left.get(prev)[1] <= left.get(i)[1]){
                result += Math.abs(left.get(prev)[0] - left.get(i)[0]) * left.get(prev)[1];
                prev = i;
            }
        }

        prev = 0;
        for (int i = 1; i < right.size(); i++) {
            if (right.get(prev)[1] <= right.get(i)[1]){
                result += Math.abs(right.get(prev)[0] - right.get(i)[0]) * right.get(prev)[1];
                prev = i;
            }
        }

        System.out.println(result + mx);
    }
}

